src/http.cc

```diff
@@ -1754,7 +1869,8 @@ httpFixupAuthentication(HttpRequest * request, const HttpHeader * hdr_in, HttpHe
  * note: initialised the HttpHeader, the caller is responsible for Clean()-ing
  */
 void
-HttpStateData::httpBuildRequestHeader(HttpRequest * request,
+HttpStateData::httpBuildRequestHeader(HttpStateData *state,
```

Comment:
I see the following problem: the code below within `mkUpgradeHeader()` assumes that `state` is non-nil, which is wrong for htcp and tunnel code. We could add checks there of course, but it seems to me that having a method parameter which does not have a sense in most cases (i.e., is nil) is not good. We need to pass `HttpStateData` there in order to initialize `HttpStateData::upgradeProtocols`.

Please consider the following refactoring proposal:
* In `upgradeProtocols` field, store a shared_ptr of a raw pointer std::vector<SBuf> *. We would not care about deletion in the destructor then.
* Remove `HttpStateData*` parameter from `mkUpgradeHeader()`, create and return the vector of upgrades as a shared_ptr.
* Remove `HttpStateData*` parameter from httpBuildRequestHeader() and return that shared_ptr. In most contexts, it would contain nil and callers would just ignore it. Only `HttpStateData::buildRequestPrefix()` would use that returned value: `upgradeProtocols = httpBuildRequestHeader(request.getRaw(), entry, fwd->al, &hdr, flags);`

Reply:
The httpBuildRequestHeader should be a non static member of HttpStateData...
In the other hand passing an HttpStateData object in an HttpStateData static member is not so bad. Anyway.

Return the upgradeProtocols list from httpBuildRequestHeader I do not like too. 

Probably the best is to add Upgrade header to hdr_out just before call httpBuildRequestHeader. The httpBuildRequestHeader will add the other headers and will check correctly for Upgrade header to add the Connection header.

Probably we need  Alex opinion for this.

Reply:
I added a check for non-NULL HttpStateData object before call mkUpgradeHeader and also an assertion check inside mkUpgradeHeader.

Addressed at 9af51fc

Reply:
> The httpBuildRequestHeader should be a non static member of HttpStateData...

We cannot do this, I think, since in htcp and tunnel code `state` is nil.

Reply:
If you can call makeUpgradeHeader() before calling httpBuildRequestHeader(), please do that. Doing so will address many of the concerns related to doing something special for HttpStateData in a function that has other callers, with other needs. I agree that httpBuildRequestHeader() and friends should be completely refactored, but I hope we do not have to do that refactoring in this mostly unrelated issue.